pkg/declared/resources.go

```diff
@@ -68,6 +68,16 @@ func (r *Resources) Update(ctx context.Context, objects []client.Object, commit
 
 	// Record the declared_resources metric, after parsing but before validation.
 	metrics.RecordDeclaredResources(ctx, commit, len(newObjects))
+	if oldCommit != commit {
+		// Reset the stream value of previous commit to 0 to provide correct input
+		// for the Otel Collector metricstransform processor's MAX aggregation.
+		// This is a temporary fix for issue b/321875474, ensuring the metrics
+		// accurately reflect decreases in declared resources. Eventually we should
+		// migrate to otel-collector-go to use the async gauge so that older streams
+		// won't get updated when there's no longer new values.
```

Comment:
Without this "reset", how long does the collector hold onto the old value and continue reporting it?

Is there another way to remove the metric, instead of setting the value to zero?

Reply:
The old values will on get cleared when the reconciler restarts, which is not regular and user needs to manually force it as of now to correct the metric data.

I haven't seen a good way to remove older streams, iiuc we'll need to retain all previous commits to be able to identify the historic streams, which would be unreasonable for some rapid source updating clients. One good way to solve this organically is as stated in comment - migrate to Otel Collector metric SDK and use the async gauge.